Fixes #38267 - rake ovirt:drop

stejskalleos opened the pull request:
Rake task for removal of Ovirt compute resources and its data
  - Disassociate all hosts from the CR
  - Update host groups
  - Remove Compute resource & its profiles

Note:
* The code removal will follow in the upcoming PR
* The rake task should work with removed Ovirt code
* This rake task is going to be cherry-picked to earlier versions, so users can run it before upgrading their instances
* It will be a migration as well.

stejskalleos committed:
Fixes #38267 - rake ovirt:drop

Rake task for removal of Ovirt compute resources and its data
  - Disassociate all hosts from the CR
  - Update host groups
  - Remove Compute resource & its profiles

stejskalleos commented:
@Lennonka text review please. This should also be documented in the upgrading workflow. The versions to chery-pick have not been decided yet.

chris1984 reviewed (commented):
@stejskalleos want me to test this?

ekohl reviewed (commented):
Without looking at the exact code I'd expect an RFC for such a user visible change as well as deprecation notices in the release notes.

> The code removal will follow in the upcoming PR

This will also need packaging changes since we have a `foreman-ovirt` subpackage.

stejskalleos commented:
> Without looking at the exact code I'd expect an RFC for such a user visible change as well as deprecation notices in the release notes.

Good point, I'll create an RFC.

I was under the impression that the deprecation was announced, but looking at the [documentation code](https://github.com/theforeman/foreman-documentation/blob/master/guides/common/modules/ref_supported-virtualization-infrastructures.adoc?plain=1), it seems like it was deprecated only for Satellite.
@Lennonka together with this PR we should announce deprecation of the Ovirt for the upcoming upstream version
The Ovirt code can then be removed in the next version.

> This will also need packaging changes since we have a foreman-ovirt subpackage.

Yop. Packaging and installer PRs will be in the scope.

stejskalleos commented:
@chris1984 QE's will be testing it before the merge, but if you have time, It will be good to have it tested by dev as well

ekohl commented on lib/tasks/ovirt.rake:
```diff
@@ -0,0 +1,66 @@
+  end
+
+  def ask_user
+    puts "!!! WARNING !!! This operation is irreversible."
```
Technically you can link them back up so I'm not sure it's really that irreversible. Just quite a bit of work.

ekohl commented on lib/tasks/ovirt.rake:
```diff
@@ -0,0 +1,66 @@
+      host_groups = cr.hostgroups.unscoped
+
+      hosts.each do |host|
+        puts "Disassociating host '#{host.name}' from the compute resource"
```
You didn't log which compute resource. Perhaps add some `puts` at the start of the loop?

ekohl commented on lib/tasks/ovirt.rake:
```diff
@@ -0,0 +1,66 @@
+  end
+
+  def the_end
+    puts 'Ovirt cleanup was successfully completed.'
```
```suggestion
    puts 'oVirt cleanup was successfully completed.'
```

ekohl commented on lib/tasks/ovirt.rake:
```diff
@@ -0,0 +1,66 @@
+
+  def ask_user
+    puts "!!! WARNING !!! This operation is irreversible."
+    puts " - This task will remove all Ovirt / RHEV compute resources from Foreman."
```
I didn't add suggestions for the code comments, but it's `oVirt`
```suggestion
    puts " - This task will remove all oVirt / RHEV compute resources from Foreman."
```

ekohl reviewed (commented):
On the rake task I'm wondering if we should somehow generalize it so it could be used for more compute resources. I have some vague recollection about users wanting to disassociate VMs from the UI, but I think that was also from before you could avoid deleting a VM when you deleted the Foreman host.

Lennonka commented on lib/tasks/ovirt.rake:
```diff
@@ -0,0 +1,66 @@
+        host.disassociate!
+      end
+
+      puts "Removing Compute Resource and Compute Profile from host groups."
```
```suggestion
      puts "Removing compute resources and compute profiles from host groups."
```

I don't see a reason for capitalization. I'd use plural instead.

Lennonka commented on lib/tasks/ovirt.rake:
```diff
@@ -0,0 +1,66 @@
+    puts " - This task will remove all Ovirt / RHEV compute resources from Foreman."
+    puts " - All hosts associated with these compute resources will be disassociated and kept."
+    puts " - All host groups with compute resources will have their compute resource and compute profile removed."
+    puts " - Compute profiles associated with the compute resources will be removed.."
```
```suggestion
    puts " - Compute profiles associated with the compute resources will be removed."
```

Lennonka commented on lib/tasks/ovirt.rake:
```diff
@@ -0,0 +1,66 @@
+namespace :ovirt do
+  desc <<~END_DESC
+    Remove all Ovirt / RHEV compute resources & profiles from Foreman.
+    All hosts associated with these compute resources will be disassociated and kept.
+    All host groups with Ovirt compute resources & profiles will have their compute resource and compute profile removed.
```
```suggestion
    Remove all oVirt / RHEV compute resources & profiles from Foreman.
    All hosts associated with these compute resources will be disassociated and kept.
    All host groups with oVirt compute resources & profiles will have their compute resource and compute profile removed.
```

Lennonka reviewed (changes requested):
Just a couple of nitpicks. + I agree with Ewoud

ekohl commented:
@ColeHiggins2 for your visibility: this is something to keep track of for Foreman 3.14.

ShimShtein commented:
@stejskalleos I think you also need to remove compute_attributes from the relevant compute profiles.

stejskalleos commented:
@ekohl @Lennonka Log messages fixed

@ShimShtein Compute Attrs should be removed together with the compute resource, see https://github.com/theforeman/foreman/blob/develop/app/models/compute_resource.rb#L28

ekohl commented on lib/tasks/ovirt.rake:
```diff
@@ -0,0 +1,65 @@
+
+    nothing_to_do if compute_resources.empty?
+
+    ask_user
```
If we're going to call this from automated scripts (foreman-maintain comes to mind) we need a way to disable this prompt.

@evgeni would you agree and if so, what's a good way to do this?

stejskalleos replied:
how about env variables, something like `SKIP_OVIRT_CHECK`

evgeni replied:
I have no idea how we did it in the past (e.g. Puppet), so probably not the right person to ask, but env var doesn't sound crazy

ekohl replied:
I did think about an env var, but something like `ASSUME_YES`. The more specific inclusion of ovirt is IMHO not a bad thing, but I wouldn't call this a check. Perhaps `OVIRT_ASSUME_YES` as a compromise?
```suggestion
    ask_user unless ENV['OVIRT_ASSUME_YES'] == '1'
```

Possibly also reuse some string to bool conversion instead of only checking for `'1'`

evgeni commented on lib/tasks/ovirt.rake:
```diff
@@ -0,0 +1,65 @@
+namespace :ovirt do
+  desc <<~END_DESC
+    Remove all oVirt / RHEV compute resources & profiles from Foreman.
```
It's been called RHV since 4.0 in 2016 ;)

evgeni commented on lib/tasks/ovirt.rake:
```diff
@@ -0,0 +1,65 @@
+    puts " - All host groups with compute resources will have their compute resource and compute profile removed."
+    puts " - Compute profiles associated with the compute resources will be removed."
+    puts ""
+    puts "Are you sure you want to continue? (yes/no)"
```
Should this print the list of affected (compute) resources?

stejskalleos commented:
* Support for `OVIRT_ASSUME_YES`
* Listing affected compute resources in the warning
* RHEV -> RHV

stejskalleos commented:
Not sure why the CI is failing but the PR is ready for re-review. cc @chris1984 @ekohl

stejskalleos commented:
And it's all green! @nofaralfasi can you review please?

ekohl commented:
> This rake task is going to be cherry-picked to earlier versions, so users can run it before upgrading their instances

To be explicit: you want this in 3.14?

nofaralfasi commented:
> On the rake task I'm wondering if we should somehow generalize it so it could be used for more compute resources. 

Why would a user want to remove all compute resources of a provider while keeping the hosts? Is that a valid use case?
In general, I agree that we should make it more flexible. However, perhaps we should consider a different RFC, like one for completely removing a compute resource - which would also delete all associated hosts and hostgroups.
Alternatively, we could add an option that prompts the user to decide if they want to remove the hosts and hostgroups as well (maybe something we should consider for this PR as well).

ekohl commented:
> Why would a user want to remove all compute resources of a provider while keeping the hosts? Is that a valid use case?

Depends on their migration plan. This way they don't need to migrate their hosts, but they can no longer do things like power management or delete the actual VM. IMHO that's valid. It may also be that they will bulk migrate them to some other platform and don't want to import them again. Instead, they can (if Foreman supports the compute resource) link them back up to the new CR.

> In general, I agree that we should make it more flexible. However, perhaps we should consider a different RFC, like one for completely removing a compute resource - which would also delete all associated hosts and hostgroups.

I think that's dangerous. Deleting hosts will terrify any sysadmin. Worst case you wipe out an entire infrastructure. Hostgroups may also be mixed.

> Alternatively, we could add an option that prompts the user to decide if they want to remove the hosts and hostgroups as well (maybe something we should consider for this PR as well).

A noop mode is generally speaking a good idea, listing what would be affected.

stejskalleos commented:
> To be explicit: you want this in 3.14?

It would be nice to have it there, as it gives users more time for migration and preparation. But it's not a requirement.

> like one for completely removing a compute resource

I want to leave that up to the user. As @ekohl said, removing hosts and host groups might be terrifying.

> A noop mode is generally speaking a good idea, listing what would be affected.

I'll add it; that's a good idea.

stejskalleos commented:
Updated:

* Added `pretend mode` for listing affected resources.
* Using `User.as_anonymous_admin` instead of `.unscoped`
* Improved description and help message.

stejskalleos commented:
And it's :green_apple: 
@nofaralfasi review please
@amolpati30 ready for QE

amolpati30 commented:
@stejskalleos 
Tested with this patch, and it successfully removes the Compute Resource, Compute Profile, and RHV from the Host Group. However, testing for the Host was not possible as the manifest is not working on stream.

stejskalleos commented:
@nofaralfasi any comments, or can we merge?

nofaralfasi commented on lib/tasks/ovirt.rake:
```diff
@@ -0,0 +1,122 @@
+    end
+  end
+
+  def ask_user(compute_resources)
```
```suggestion
  def prompt_user_confirmation(compute_resources)
```

nofaralfasi replied:
I still think we should use a more informative name method here.

nofaralfasi commented on lib/tasks/ovirt.rake:
```diff
@@ -0,0 +1,122 @@
+  end
+
+  def the_end
+    puts ''
+    puts 'oVirt cleanup was successfully completed.'
```
Why not use `\n` within a single `puts` instead of multiple `puts` calls?
```suggestion
    puts '\noVirt cleanup was successfully completed.'
```

nofaralfasi commented on lib/tasks/ovirt.rake:
```diff
@@ -0,0 +1,122 @@
+    Compute profiles, images, and other resources will be removed.
+
+    Use OVIRT_PRETEND=true to simulate the operation without making any changes.
+    Use OVIRT_ASSUME_YES=true to skip the confirmation prompt.
```
How about using a more descriptive name, like `OVIRT_SKIP_CONFIRM` or `OVIRT_FORCE`?

nofaralfasi commented on lib/tasks/ovirt.rake:
```diff
@@ -0,0 +1,122 @@
+    the_end
+  end
+
+  def run_in_pretend_mode(compute_resources)
```
```suggestion
  def simulate_cleanup(compute_resources)
```

nofaralfasi commented on lib/tasks/ovirt.rake:
```diff
@@ -0,0 +1,122 @@
+    exit 0
+  end
+
+  def run_for_real(compute_resources)
```
```suggestion
  def perform_cleanup(compute_resources)
```

nofaralfasi commented on lib/tasks/ovirt.rake:
```diff
@@ -0,0 +1,122 @@
+    end
+  end
+
+  def nothing_to_do
```
```suggestion
  def exit_if_no_resources
```

nofaralfasi commented on lib/tasks/ovirt.rake:
```diff
@@ -0,0 +1,122 @@
+    exit 0
+  end
+
+  def the_end
```
```suggestion
  def finalize_cleanup
```

nofaralfasi reviewed (changes requested):
Another suggestion: You have multiple methods that print a message and exit. Why not refactor them into a single method that takes a message and an exit status as arguments?

stejskalleos commented:
> Another suggestion: You have multiple methods that print a message and exit. Why not refactor them into a single method that takes a message and an exit status as arguments?

Most of them have multi-line output, and storing that output in a variable and then passing it to the method is not very readable, IMO. So, in this case, I chose readability over DRY.

nofaralfasi reviewed (dismissed):
I left all my notes and let Leos decide which ones to address. Other than that, the code looks good to me, but I didn’t test it.

stejskalleos commented:
@nofaralfasi Sorry, my bad, I accidentally missed the `prompt_user_confirmation` rename. Now it should be 1:1 with your comments.

amolpati30 commented:
@stejskalleos 
I have tested and LGTM.
